File: java/search/top_k/HeapSelect.java
package search.top_k;

import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.Queue;

public class HeapSelect {
    public int[] search(int[] nums, int k) {
        Queue<Integer> queue = new PriorityQueue<>(cmp);
        for (int value : nums) {
            if (queue.size() < k) {
                queue.offer(value);
                continue;
            }

            if (value < queue.peek()) {
                queue.poll();
                queue.offer(value);
            }
        }
        return queue.stream().mapToInt(Integer::valueOf).toArray();
    }

    private static Comparator<Integer> cmp = new Comparator<Integer>() {
        @Override
        public int compare(Integer i, Integer j) {
            return j - i;
        }
    };
}

File: cpp/search/top_k/HeapSelect.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

class HeapSelect {
public:
    vector<int> search(vector<int> &nums, int k) {
        priority_queue<int, vector<int>, cmp> queue;
        for (int value : nums) {
            if ((int) queue.size() < k) {
                queue.push(value);
                continue;
            }

            if (value < queue.top()) {
                queue.pop();
                queue.push(value);
            }
        }

        vector<int> result(0);
        while (!queue.empty()) {
            result.emplace_back(queue.top());
            queue.pop();
        }
        return result;
    }

private:
    struct cmp {
        bool operator() (const int p, const int q) {
            return p < q;
        }
    };
};
